File: py/randmove.py

import go
import random


class Policy():
    def __init__(self, PUCT=1):
        self.PUCT = PUCT
        self.HASH = {}

    def get(self, position):
        positions = position.get_children()

        random.shuffle(positions)
        positions.insert(0, position.move(0))

        return positions

    def sim(self, position):
        score = self.HASH.get(position.hash_code)
        if score is not None:
            return score

        pos = position

        while pos.pass_count() < 2:
            positions = pos.get_children()
            n = len(positions)
            if n >= 2:
                i = random.randint(0, n-1)
                pos = positions.pop(i)
                for p in positions:
                    p.release()
            elif n == 1:
                pos = positions[0]
            else:
                pos = pos.move(0)

        score = pos.score()
        while pos is not position:
            pos.release()
            pos = pos.parent
        
        self.HASH[position.hash_code] = score

        return score

    def clear(self):
        self.HASH = {}

File: cpp/src/RandMove.cpp
#include <algorithm>
#include <chrono>
#include "RandMove.h"
#include "go.h"

using namespace std;

RandMove::RandMove(float puct) : Policy(puct)
{
}

void RandMove::get(Position *position, vector<Position *> &positions)
{
    position->getChildren(positions);
    for (auto i = positions.size()-1; i > 1; --i)
    {
        swap(positions[i], positions[rand() % i + 1]);
    }
}

float RandMove::sim(Position *position)
{
    auto it = _hash.find(position->hashCode());
    if (it != _hash.end())
    {
        return it->second;
    }

    Position *pos = position;

    vector<Position *> positions(go::LN);

    while (pos->passCount() < 2)
    {
        positions.clear();

        pos->getChildren(positions);

        size_t n = positions.size();

        Position *pp = nullptr;

        if (n >= 2)
        {
            int i = rand() % n;
            pp = positions[i];
            for (auto it = positions.begin(); it != positions.end(); ++it)
            {
                Position *p = *it;
                if (p != pp)
                {
                    p->release();
                }
            }
        }
        else if (n == 1)
        {
            pp = positions[0];
        }
        else
        {
            pp = pos->move(go::PASS);
        }

        pos = pp;
    }

    float score = pos->score();
    while (pos != position)
    {
        pos->release();
        pos = pos->parent();
    }

    _hash.insert(make_pair(position->hashCode(), score));

    return score;
}

void RandMove::clear(void)
{
    _hash.clear();
}